### ISSUE TITLE
Bad state: Can't finalize a finalized MultipartFile

### ISSUE THREAD

**Issue by claudiowagner** [flutterchina/dio#482]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                       |
| ------------------------------ | ----------------------------------------------------- |
| Platform Name                  | android                                |
| Platform Version               | 9.0                                 |
| Dio Version                    | 3.0.1                                        |
| Android Studio | Android Studio 3.5                   |
| Repro rate                     | all the time (100%)         |
| Repro with our demo prj        | no                |
| Demo project link              | no      |

### Flutter
[√] Flutter (Channel stable, v1.9.1+hotfix.2, on Microsoft Windows [versÃ£o 10.0.17763.737], locale pt-BR)
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.5)
[√] Connected device (1 available)

### Issue Description and Steps

The update token example is not working if repeat is performed with a Post / FormData.

1- Use an interceptor to update token, as in the [interceptor_lock](https://github.com/flutterchina/dio/blob/master/example/interceptor_lock.dart) example;
2- Run a POST/FormData;
3- After updating the token, the request is resubmitted;

```
I/flutter ( 9472): DioError [DioErrorType.DEFAULT]: Bad state: Can't finalize a finalized MultipartFile.
I/flutter ( 9472): #0      FormData.finalize (package:dio/src/form_data.dart:124:7)
I/flutter ( 9472): #1      DioMixin._transformData (package:dio/src/dio.dart:987:23)
I/flutter ( 9472): <asynchronous suspension>
I/flutter ( 9472): #2      DioMixin._dispatchRequest (package:dio/src/dio.dart:897:26)
I/flutter ( 9472): <asynchronous suspension>
I/flutter ( 9472): #3      DioMixin._request._interceptorWrapper.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:dio/src/dio.dart:828:37)
I/flutter ( 9472): #4      DioMixin.checkIfNeedEnqueue (package:dio/src/dio.dart:1100:22)
I/flutter ( 9472): #5      DioMixin._request._interceptorWrapper.<anonymous closure>.<anonymous closure> (package:dio/src/dio.dart:825:22)
I/flutter ( 9472): #6      new Future.<anonymous closure> (dart:async/future.dart:176:37)
I/flutter ( 9472): #7      _rootRun (dart:async/zone.dart:1120:38)
I/flutter ( 9472): #8      _CustomZone.run (dart:async/zone.dart:1021:19)
I/flutter ( 9472): #9      _CustomZone.runGuarded (dart:async/zone.dart:923:7)
I/flutter ( 9472): #10     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
I/flutter ( 9472): #11     _rootRun (dart:async/zo
I/flutter ( 9472): #0      DioMixin._request._errorInterceptorWrapper.<anonymous closure> (package:dio/src/dio.dart:848:13)
I/flutter ( 9472): <asynchronous suspension>
I/flutter ( 9472): #1      _rootRunUnary (dart:async/zone.dart:1132:38)
I/flutter ( 9472): #2      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
I/flutter ( 9472): #3      _FutureListener.handleError (dart:async/future_impl.dart:155:20)
I/flutter ( 9472): #4      Future._propagateToListeners.handleError (dart:async/future_impl.dart:690:47)
I/flutter ( 9472): #5      Future._propagateToListeners (dart:async/future_impl.dart:711:24)
I/flutter ( 9472): #6      Future._completeError (dart:async/future_impl.dart:530:5)
I/flutter ( 9472): #7      _SyncCompleter._completeError (dart:async/future_impl.dart:55:12)
I/flutter ( 9472): #8      _Completer.completeError (dart:async/future_impl.dart:27:5)
I/flutter ( 9472): #9      Future.any.<anonymous closure> (dart:async/future.dart:462:45)
I/flutter ( 9472): #10     _rootRunBinary (dart:async/zone.dart:1144:38)
I/flutter ( 9472): #11     _CustomZone.runBinary (dart:async/zone.dart:1037:19)
I/flutter ( 9472): #12     _FutureListener.handleError (dart:async/future_impl.dart:151:20)
I/flutter ( 9472): #13     Future._propagateToListeners.handleError (dart:async/future_impl.dart:6
```

**Comment by wendux** [comment]
Because ` MultipartFile` is based on `Stream`, and a `Stream` can be read only  once, you should create a new ` MultipartFile` when the request is resubmitted.

**Comment by claudiowagner** [comment]
> Because ` MultipartFile` is based on `Stream`, and a `Stream` can be read only once, you should create a new ` MultipartFile` when the request is resubmitted.

I tried this way, but the original `MultipartFile` does not have the file path to create a new `FormData`.

```dart
// The rest is the same of interceptor_lock example
}).then((e) {
  //repeat
  if (options.data is FormData) {
	FormData formData = FormData();
	formData.fields.addAll(options.data.fields);

	for (MapEntry mapFile in options.data.files) {
	  formData.files.add(MapEntry(
		  mapFile.key,
		  MultipartFile.fromFileSync(???mapFile.value.FILE_PATH???,
			  filename: mapFile.value.filename)));
	}
	options.data = formData;
  }
  return dio.request(options.path, options: options);
}
```

**Comment by ramonweiss** [comment]
I have the same issue, any tips?

**Comment by AllenWen** [comment]
> > Because ` MultipartFile` is based on `Stream`, and a `Stream` can be read only once, you should create a new ` MultipartFile` when the request is resubmitted.
> 
> I tried this way, but the original `MultipartFile` does not have the file path to create a new `FormData`.
> 
> ```dart
> // The rest is the same of interceptor_lock example
> }).then((e) {
>   //repeat
>   if (options.data is FormData) {
> 	FormData formData = FormData();
> 	formData.fields.addAll(options.data.fields);
> 
> 	for (MapEntry mapFile in options.data.files) {
> 	  formData.files.add(MapEntry(
> 		  mapFile.key,
> 		  MultipartFile.fromFileSync(???mapFile.value.FILE_PATH???,
> 			  filename: mapFile.value.filename)));
> 	}
> 	options.data = formData;
>   }
>   return dio.request(options.path, options: options);
> }
> ```

thx!saved my day!

**Comment by xuzhengquan** [comment]
> > > Because ` MultipartFile` is based on `Stream`, and a `Stream` can be read only once, you should create a new ` MultipartFile` when the request is resubmitted.
> > 
> > 
> > I tried this way, but the original `MultipartFile` does not have the file path to create a new `FormData`.
> > ```dart
> > // The rest is the same of interceptor_lock example
> > }).then((e) {
> >   //repeat
> >   if (options.data is FormData) {
> > 	FormData formData = FormData();
> > 	formData.fields.addAll(options.data.fields);
> > 
> > 	for (MapEntry mapFile in options.data.files) {
> > 	  formData.files.add(MapEntry(
> > 		  mapFile.key,
> > 		  MultipartFile.fromFileSync(???mapFile.value.FILE_PATH???,
> > 			  filename: mapFile.value.filename)));
> > 	}
> > 	options.data = formData;
> >   }
> >   return dio.request(options.path, options: options);
> > }
> > ```
> 
> thx!saved my day!

how to get the multipartFile file path ???

**Comment by AllenWen** [comment]
> > Because ` MultipartFile` is based on `Stream`, and a `Stream` can be read only once, you should create a new ` MultipartFile` when the request is resubmitted.
> 
> I tried this way, but the original `MultipartFile` does not have the file path to create a new `FormData`.
> 
> ```dart
> // The rest is the same of interceptor_lock example
> }).then((e) {
>   //repeat
>   if (options.data is FormData) {
> 	FormData formData = FormData();
> 	formData.fields.addAll(options.data.fields);
> 
> 	for (MapEntry mapFile in options.data.files) {
> 	  formData.files.add(MapEntry(
> 		  mapFile.key,
> 		  MultipartFile.fromFileSync(???mapFile.value.FILE_PATH???,
> 			  filename: mapFile.value.filename)));
> 	}
> 	options.data = formData;
>   }
>   return dio.request(options.path, options: options);
> }
> ```

I found this solution will convert form into json，seems not right...

**Comment by matanshukry** [comment]
Why has this been issue closed? it's still relevant today.

# Possible solutions

1. Add a class derived from `MultipartFile` and has an extra field `filePath`.
2. Add a field to `MultipartFile` class called `extra`, much like the one in `Options`.

# Workaround

If anyone is looking for a workaround, one can:
1. Use option 1 in possible solution, but create the derived class your self. When retrying the operation, check if the multipart file is of your derived class, and get the path if so.
2. Use the `Options.extra`  field and pass a map between the keys in `FormData` to a path

**Comment by liweiwh** [comment]
@matanshukry  you can easily add field 'path' to your request `header` part, and use it in repeat request.

**Comment by matanshukry** [comment]
@liweiwh Sending extra data over the network when it's not needed (it's used locally only) is a bad option

**Comment by liweiwh** [comment]
@matanshukry you can delete it when you do the request.

```
      if (status.toString().startsWith('0')) {
        token = resMap['data']['token'];
        await store.setString(StoreKeys.token, token);
        _options.headers['authorization'] = token;
        dio.unlock(); 
        if (_options.data is FormData) {
          FormData formData = FormData();
          formData.fields.addAll(_options.data.fields);

          for (MapEntry mapFile in _options.data.files) {
            formData.files.add(MapEntry(
                mapFile.key,
                await MultipartFile.fromFile(_options.headers['path'],
                    filename: mapFile.value.filename)));
          }
          _options.data = formData;
        }
        // empty the path field
        _options.headers['path'] = null;
        return dio.request(path, options: _options);
      }
```

**Comment by matanshukry** [comment]
If you delete it from the request, how will it exist in the interceptor?

**Comment by liweiwh** [comment]
@matanshukry oh, sorry, my mistake, I use it in my refresh token process, but when the first time do the request the `path` is in the header. Not solve your problem, sorry.

**Comment by felixivance** [comment]
If you are resending the same formData in a nested dio request eg
   ` FormData formData = ...`
`Dio().post('url', data:formData).then((_){`
 `//re-initialise form data here`
    `Dio().post('url', data:formData).then((_){`
       `...`
   ` });`

`}).catch( ...`

be sure to re-initiate form data because it is already finalized

**Comment by indokan2001** [comment]
I made my own custom multipart class that can store file path, if anyone is interested:

```dart
import 'dart:async';
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:http_parser/http_parser.dart';
import 'package:path/path.dart' as p;

class MultipartFileExtended extends MultipartFile {
  final String filePath;  //this one!

  MultipartFileExtended(
    Stream<List<int>> stream,
    length, {
    filename,
    this.filePath,
    contentType,
  }) : super(stream, length, filename: filename, contentType: contentType);

  static MultipartFileExtended fromFileSync(
    String filePath, {
    String filename,
    MediaType contentType,
  }) => multipartFileFromPathSync(filePath, filename: filename, contentType: contentType);
}


MultipartFileExtended multipartFileFromPathSync(
  String filePath, {
  String filename,
  MediaType contentType,
}) {
  filename ??= p.basename(filePath);
  var file = File(filePath);
  var length = file.lengthSync();
  var stream = file.openRead();
  return MultipartFileExtended(
    stream,
    length,
    filename: filename,
    contentType: contentType,
    filePath: filePath,
  );
}
```
Now you can call request like this:
```dart
Dio dio = new Dio();
FormData formData = FormData.fromMap({
  "file": MultipartFileExtended.fromFileSync("./example/upload.txt",
          filename: "upload.txt")
});

var response = await dio.post("/info", data: formData).then((e) {
  FormData newFormData = FormData();
  newFormData.fields.addAll(formData.fields);

  for (MapEntry mapFile in formData.files) {
    newFormData.files.add(MapEntry(
        mapFile.key,
        MultipartFileExtended.fromFileSync(mapFile.value.filePath, //fixed!
            filename: mapFile.value.filename)));
  }

  return dio.post("/info", data: newFormData);
});
```

**Comment by dotw** [comment]
thanks @indokan2001 , your solution works for my case.

**Comment by Refaat-taha** [comment]
the problem still exists  
please, anyone, have a complete solution  please share it with me

**Comment by AbdOoSaed** [comment]
after I review my code found
by mistake, I send the same Multipart file twice 

I know this not closely related to the issue, but I think my comment will help someone
![temp](https://user-images.githubusercontent.com/33700292/124502421-4f3ace00-dd78-11eb-82bc-19babc7768d4.PNG)

**Comment by befirst** [comment]
just create a function with dio request and use it again when you came across this error

**Comment by mosi2142** [comment]
> ```dart
> newFormData
> ```

i get this error 

> 
> 
> I made my own custom multipart class that can store file path, the solution as @matanshukry pointed out, if anyone is interested:
> 
> ```dart
> import 'dart:async';
> import 'dart:io';
> import 'package:dio/dio.dart';
> import 'package:http_parser/http_parser.dart';
> import 'package:path/path.dart' as p;
> 
> class MultipartFileExtended extends MultipartFile {
>   final String filePath;  //this one!
> 
>   MultipartFileExtended(
>     Stream<List<int>> stream,
>     length, {
>     filename,
>     this.filePath,
>     contentType,
>   }) : super(stream, length, filename: filename, contentType: contentType);
> 
>   static MultipartFileExtended fromFileSync(
>     String filePath, {
>     String filename,
>     MediaType contentType,
>   }) => multipartFileFromPathSync(filePath, filename: filename, contentType: contentType);
> }
> 
> 
> MultipartFileExtended multipartFileFromPathSync(
>   String filePath, {
>   String filename,
>   MediaType contentType,
> }) {
>   filename ??= p.basename(filePath);
>   var file = File(filePath);
>   var length = file.lengthSync();
>   var stream = file.openRead();
>   return MultipartFileExtended(
>     stream,
>     length,
>     filename: filename,
>     contentType: contentType,
>     filePath: filePath,
>   );
> }
> ```
> 
> Now you can call request like this:
> 
> ```dart
> Dio dio = new Dio();
> FormData formData = FormData.fromMap({
>   "file": MultipartFileExtended.fromFileSync("./example/upload.txt",
>           filename: "upload.txt")
> });
> 
> var response = await dio.post("/info", data: formData).then((e) {
>   FormData newFormData = FormData();
>   newFormData.fields.addAll(formData.fields);
> 
>   for (MapEntry mapFile in formData.files) {
>     newFormData.files.add(MapEntry(
>         mapFile.key,
>         MultipartFileExtended.fromFileSync(mapFile.value.filePath, //fixed!
>             filename: mapFile.value.filename)));
>   }
> 
>   return dio.post("/info", data: newFormData);
> });
> ```

i get this error when using the class 
`Unhandled Exception: NoSuchMethodError: Class 'MultipartFile' has no instance getter 'filePath'.
E/flutter ( 5189): Receiver: Instance of 'MultipartFile'`

**Comment by mosi2142** [comment]
> 
> 
> I made my own custom multipart class that can store file path, the solution as @matanshukry pointed out, if anyone is interested:
> 
> ```dart
> import 'dart:async';
> import 'dart:io';
> import 'package:dio/dio.dart';
> import 'package:http_parser/http_parser.dart';
> import 'package:path/path.dart' as p;
> 
> class MultipartFileExtended extends MultipartFile {
>   final String filePath;  //this one!
> 
>   MultipartFileExtended(
>     Stream<List<int>> stream,
>     length, {
>     filename,
>     this.filePath,
>     contentType,
>   }) : super(stream, length, filename: filename, contentType: contentType);
> 
>   static MultipartFileExtended fromFileSync(
>     String filePath, {
>     String filename,
>     MediaType contentType,
>   }) => multipartFileFromPathSync(filePath, filename: filename, contentType: contentType);
> }
> 
> 
> MultipartFileExtended multipartFileFromPathSync(
>   String filePath, {
>   String filename,
>   MediaType contentType,
> }) {
>   filename ??= p.basename(filePath);
>   var file = File(filePath);
>   var length = file.lengthSync();
>   var stream = file.openRead();
>   return MultipartFileExtended(
>     stream,
>     length,
>     filename: filename,
>     contentType: contentType,
>     filePath: filePath,
>   );
> }
> ```
> 
> Now you can call request like this:
> 
> ```dart
> Dio dio = new Dio();
> FormData formData = FormData.fromMap({
>   "file": MultipartFileExtended.fromFileSync("./example/upload.txt",
>           filename: "upload.txt")
> });
> 
> var response = await dio.post("/info", data: formData).then((e) {
>   FormData newFormData = FormData();
>   newFormData.fields.addAll(formData.fields);
> 
>   for (MapEntry mapFile in formData.files) {
>     newFormData.files.add(MapEntry(
>         mapFile.key,
>         MultipartFileExtended.fromFileSync(mapFile.value.filePath, //fixed!
>             filename: mapFile.value.filename)));
>   }
> 
>   return dio.post("/info", data: newFormData);
> });
> ```
@hmdfthid 
i get this error when using this custom class 

`Unhandled Exception: NoSuchMethodError: Class 'MultipartFile' has no instance getter 'filePath'.
E/flutter ( 5189): Receiver: Instance of 'MultipartFile'`

**Comment by chuanpham** [comment]
@mosi2142 did you solve the problem?

**Comment by akj-seq** [comment]
just replace file path  "/" with a "@" and make it as a filename in multipartfile and after refresh token resending request use this filename and restore it by replacing "@" with "/" now path is ready create a new multipartfile with that path. solved!!.

**Comment by jeankpoti** [comment]
You can just reset the list in which you are storing the multipartfiles after the request is made.

**Comment by MingSern** [comment]
> I made my own custom multipart class that can store file path, the solution as @matanshukry pointed out, if anyone is interested:
> 
> ```dart
> import 'dart:async';
> import 'dart:io';
> import 'package:dio/dio.dart';
> import 'package:http_parser/http_parser.dart';
> import 'package:path/path.dart' as p;
> 
> class MultipartFileExtended extends MultipartFile {
>   final String filePath;  //this one!
> 
>   MultipartFileExtended(
>     Stream<List<int>> stream,
>     length, {
>     filename,
>     this.filePath,
>     contentType,
>   }) : super(stream, length, filename: filename, contentType: contentType);
> 
>   static MultipartFileExtended fromFileSync(
>     String filePath, {
>     String filename,
>     MediaType contentType,
>   }) => multipartFileFromPathSync(filePath, filename: filename, contentType: contentType);
> }
> 
> 
> MultipartFileExtended multipartFileFromPathSync(
>   String filePath, {
>   String filename,
>   MediaType contentType,
> }) {
>   filename ??= p.basename(filePath);
>   var file = File(filePath);
>   var length = file.lengthSync();
>   var stream = file.openRead();
>   return MultipartFileExtended(
>     stream,
>     length,
>     filename: filename,
>     contentType: contentType,
>     filePath: filePath,
>   );
> }
> ```
> 
> Now you can call request like this:
> 
> ```dart
> Dio dio = new Dio();
> FormData formData = FormData.fromMap({
>   "file": MultipartFileExtended.fromFileSync("./example/upload.txt",
>           filename: "upload.txt")
> });
> 
> var response = await dio.post("/info", data: formData).then((e) {
>   FormData newFormData = FormData();
>   newFormData.fields.addAll(formData.fields);
> 
>   for (MapEntry mapFile in formData.files) {
>     newFormData.files.add(MapEntry(
>         mapFile.key,
>         MultipartFileExtended.fromFileSync(mapFile.value.filePath, //fixed!
>             filename: mapFile.value.filename)));
>   }
> 
>   return dio.post("/info", data: newFormData);
> });
> ```

How about MultipartFile that are from bytes?

**Comment by aweiand** [comment]
Anyone using dio_smart_retry and have the same issue? I'm using it with dio , and when get second or third retry, causes this error....

**Comment by fasterlars** [comment]
> Anyone using dio_smart_retry and have the same issue? I'm using it with dio , and when get second or third retry, causes this error....

I am using dio_smart_retry and got this error. But on the initial request, and the subsequent. 
But only got it once on a Motorola device I do not own, and I cannot reproduce it.

**Comment by vimaxwell** [comment]
@aweiand @fasterlars please test my fix https://github.com/rodion-m/dio_smart_retry/commit/ac4c291e365723115458b28f0f1cec9b5100d368

**Comment by fasterlars** [comment]
@vimaxwell I have not seen this error in a while, but will try your fix if I see it again and are able to reproduce it.
Thanks 👍

**Comment by aweiand** [comment]
Hi @vimaxwell ,

After test in a test server I own, the issue persists...

**Comment by rodion-m** [comment]
> Anyone using dio_smart_retry and have the same issue? I'm using it with dio , and when get second or third retry, causes this error....

Starting from version 1.4.0 [dio_smart_retry](https://pub.dev/packages/dio_smart_retry) package supports retrying for requests with multipart/form-data.
Just use the `MultipartFileRecreatable` class instead of `MultipartFile` to make retry available. Details: https://github.com/rodion-m/dio_smart_retry#retry-requests-with-multipartform-data

**Comment by rinlv** [comment]
> > Anyone using dio_smart_retry and have the same issue? I'm using it with dio , and when get second or third retry, causes this error....
> 
> Starting from version 1.4.0 [dio_smart_retry](https://pub.dev/packages/dio_smart_retry) package supports retrying for requests with multipart/form-data. Just use the `MultipartFileRecreatable` class instead of `MultipartFile` to make retry available. Details: https://github.com/rodion-m/dio_smart_retry#retry-requests-with-multipartform-data

Hi @rodion-m, how to `MultipartFileRecreatable` work with [retrofit](https://github.com/trevorwang/retrofit.dart/)?

**Comment by CodeFoxLk** [comment]
> > I made my own custom multipart class that can store file path, the solution as @matanshukry pointed out, if anyone is interested:
> > ```dart
> > import 'dart:async';
> > import 'dart:io';
> > import 'package:dio/dio.dart';
> > import 'package:http_parser/http_parser.dart';
> > import 'package:path/path.dart' as p;
> > 
> > class MultipartFileExtended extends MultipartFile {
> >   final String filePath;  //this one!
> > 
> >   MultipartFileExtended(
> >     Stream<List<int>> stream,
> >     length, {
> >     filename,
> >     this.filePath,
> >     contentType,
> >   }) : super(stream, length, filename: filename, contentType: contentType);
> > 
> >   static MultipartFileExtended fromFileSync(
> >     String filePath, {
> >     String filename,
> >     MediaType contentType,
> >   }) => multipartFileFromPathSync(filePath, filename: filename, contentType: contentType);
> > }
> > 
> > 
> > MultipartFileExtended multipartFileFromPathSync(
> >   String filePath, {
> >   String filename,
> >   MediaType contentType,
> > }) {
> >   filename ??= p.basename(filePath);
> >   var file = File(filePath);
> >   var length = file.lengthSync();
> >   var stream = file.openRead();
> >   return MultipartFileExtended(
> >     stream,
> >     length,
> >     filename: filename,
> >     contentType: contentType,
> >     filePath: filePath,
> >   );
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Now you can call request like this:
> > ```dart
> > Dio dio = new Dio();
> > FormData formData = FormData.fromMap({
> >   "file": MultipartFileExtended.fromFileSync("./example/upload.txt",
> >           filename: "upload.txt")
> > });
> > 
> > var response = await dio.post("/info", data: formData).then((e) {
> >   FormData newFormData = FormData();
> >   newFormData.fields.addAll(formData.fields);
> > 
> >   for (MapEntry mapFile in formData.files) {
> >     newFormData.files.add(MapEntry(
> >         mapFile.key,
> >         MultipartFileExtended.fromFileSync(mapFile.value.filePath, //fixed!
> >             filename: mapFile.value.filename)));
> >   }
> > 
> >   return dio.post("/info", data: newFormData);
> > });
> > ```
> 
> How about MultipartFile that are from bytes?

```
class MultipartFileExtended extends MultipartFile {
  MultipartFileExtended(
    Stream<List<int>> stream,
    int length, {
    String? filename,
    Map<String, List<String>>? headers,
    MediaType? contentType,
    List<int>? byteValue,
  })  : _byteValue = byteValue,
        super(stream, length, filename: filename, contentType: contentType);

  List<int>? _byteValue;
  List<int>? get byteValue => _byteValue;

  factory MultipartFileExtended.fromBytes(
    List<int> value, {
    String? filename,
    MediaType? contentType,
    final Map<String, List<String>>? headers,
  }) {
    final stream = Stream.fromIterable([value]);

    return MultipartFileExtended(stream, value.length,
        filename: filename, contentType: contentType, headers: headers, byteValue: value);
  }
}
```

**Comment by MovingMelody** [comment]
> > Because ` MultipartFile` is based on `Stream`, and a `Stream` can be read only once, you should create a new ` MultipartFile` when the request is resubmitted.
> 
> I tried this way, but the original `MultipartFile` does not have the file path to create a new `FormData`.
> 
> ```dart
> // The rest is the same of interceptor_lock example
> }).then((e) {
>   //repeat
>   if (options.data is FormData) {
> 	FormData formData = FormData();
> 	formData.fields.addAll(options.data.fields);
> 
> 	for (MapEntry mapFile in options.data.files) {
> 	  formData.files.add(MapEntry(
> 		  mapFile.key,
> 		  MultipartFile.fromFileSync(???mapFile.value.FILE_PATH???,
> 			  filename: mapFile.value.filename)));
> 	}
> 	options.data = formData;
>   }
>   return dio.request(options.path, options: options);
> }
> ```

Awesome. It worked

```dart
...
if (options.data is FormData) {
            FormData formData = FormData();
            formData.fields.addAll(options.data.fields);

            for (MapEntry mapFile in options.data.files) {
              formData.files.add(MapEntry(
                mapFile.key,
                await MultipartFile.fromFile(
                    (mapFile.value as MultipartFile).filename ?? "",
                    filename: "${(mapFile.value as MultipartFile).filename}"),
              ));
            }
            options.data = formData;
            final Response response = await dio.fetch(options);
            return handler.resolve(response);
          }
```

**Comment by gabrielaraujoz** [comment]
Hi there! This PR should solve this problem without extra workarounds. Hope you enjoy it <3

**Comment by YeeWan0217** [comment]
> if (options.data is FormData) {
>     FormData formData = FormData();
> 
>     for (MapEntry mapFile in options.data.files) {
>       formData.files.add(MapEntry(
>         mapFile.key,
>         mapFile.value.clone(),
>       ));
>     }
>     options.data = formData;
>   }

**Comment by hosseinhp1378** [comment]
> Take a long time to find, this work for me. The type of mapFile.value is MultipartFile which can use clone() method from MultipartFile Class Library
> 
> ```
> if (options.data is FormData) {
>     FormData formData = FormData();
> 
>     for (MapEntry mapFile in options.data.files) {
>       formData.files.add(MapEntry(
>         mapFile.key,
>         mapFile.value.clone(),
>       ));
>     }
>     options.data = formData;
>   }
> ```

this worked perfectly
Thank you

**Comment by AlexV525** [comment]
https://github.com/cfug/dio/blob/ff4eb26f758f665bcda9752ed60304552bc32903/dio/lib/src/multipart_file.dart#L169-L179
https://github.com/cfug/dio/blob/ff4eb26f758f665bcda9752ed60304552bc32903/dio/lib/src/form_data.dart#L210-L218
